Make WebPermissionController independent of the WebPage

RupinMittal opened the pull request:
#### 2a9bcbc919a9fa687506d51541efb66a80ee52f2
<pre>
Make WebPermissionController independent of the WebPage
<a href="https://bugs.webkit.org/show_bug.cgi?id=243597">https://bugs.webkit.org/show_bug.cgi?id=243597</a>

Reviewed by Chris Dumez and Sihui Liu.

Currently, the Permissions API is not supported for workers. In order for it
to be supported, the Permissions API should function without the need for a
page. Previously, WebPermissionController was created and owned by WebPage.
This patch makes PermissionController a singleton that is created by the
WebProcess. So now, instead of Permissions::query() using the page to get the
WebPermissionController and send an IPC message to WebPageProxy to obtain
permissions states, it can access the WebPermissionController singleton to
send an IPC message that will eventually call the correct WebPageProxy. Note
that a page is still required to obtain the permission state, but this
refactoring is the first step in supporting the Permissions API for workers.

* Source/WebCore/Modules/permissions/PermissionController.cpp: Copied from Source/WebCore/Modules/permissions/Permissions.h.
(WebCore::sharedController):
(WebCore::PermissionController::shared):
(WebCore::PermissionController::setSharedController):
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/PermissionStatus.cpp:
(WebCore::PermissionStatus::PermissionStatus):
(WebCore::PermissionStatus::~PermissionStatus):
* Source/WebCore/Modules/permissions/PermissionStatus.h:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::Permissions):
(WebCore::Permissions::query):
* Source/WebCore/Modules/permissions/Permissions.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::permissionController): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::permissionController): Deleted.
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::permissionController): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp: Copied from Source/WebCore/Modules/permissions/Permissions.h.
(WebKit::WebPermissionControllerProxy::WebPermissionControllerProxy):
(WebKit::WebPermissionControllerProxy::~WebPermissionControllerProxy):
(WebKit::WebPermissionControllerProxy::query):
* Source/WebKit/UIProcess/WebPermissionControllerProxy.h: Copied from Source/WebCore/Modules/permissions/Permissions.h.
* Source/WebKit/UIProcess/WebPermissionControllerProxy.messages.in: Copied from Source/WebCore/Modules/permissions/Permissions.h.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::m_webPermissionController):
(WebKit::m_webLockRegistry): Deleted.
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::create):
(WebKit::WebPermissionController::query):
(WebKit::WebPermissionController::tryProcessingRequests):
(WebKit::WebPermissionController::WebPermissionController): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
* Source/WebKit/WebProcess/WebProcess.cpp:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):
* Source/WebKitLegacy/win/WebView.cpp:
(WebView::initWithFrame):

Canonical link: <a href="https://commits.webkit.org/253216@main">https://commits.webkit.org/253216@main</a>
</pre>

cdumez commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "Page.h"
```
Please use a Forward declaration instead of an include.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8840,7 +8840,7 @@ void WebPageProxy::revokeGeolocationAuthorizationToken(const String& authorizati
 }
 
-void WebPageProxy::queryPermission(const ClientOrigin& clientOrigin, const PermissionDescriptor& descriptor, CompletionHandler<void(std::optional<PermissionState>, bool shouldCache)>&& completionHandler)
+void WebPageProxy::queryPermission(const ClientOrigin& clientOrigin, const PermissionDescriptor& descriptor, CompletionHandler<void(std::optional<PermissionState>, bool shouldCache)>& completionHandler)
```
Why this change? This doesn't look right.

RupinMittal replied:
I did this because WebPermissionControllerProxy takes in the completionHandler as an rvalue reference, so it when it passes it to WebPageProxy, WebPageProxy should take it in as a lvalue reference right?

cdumez replied:
You did this because you made a mistake and forgot a WTFMove(completionHandler), like I commented below.

RupinMittal replied:
Ah you're right--I see it. Fixed!

cdumez commented on Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp:
```diff
@@ -0,0 +1,59 @@
+    if (!webPageProxy)
+        completionHandler(WebCore::PermissionState::Prompt, false);
+
+    webPageProxy->queryPermission(clientOrigin, descriptor, completionHandler);
```
WTFMove(completionHandler)

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -26,33 +26,33 @@
+Ref<WebPermissionController> WebPermissionController::create()
 {
-    return adoptRef(*new WebPermissionController(page));
+    return adoptRef(*new WebPermissionController());
```
No need for `()`:
`return adoptRef(*new WebPermissionController);`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -26,33 +26,33 @@
 
-WebPermissionController::WebPermissionController(WebPage& page)
-    : m_page(page)
+WebPermissionController::WebPermissionController()
```
WebPermissionController::WebPermissionController() = default;

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -26,33 +26,33 @@
 
-    m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), WTFMove(completionHandler) });
+    auto webPageProxyIdentifier = WebPage::fromCorePage(page).webPageProxyIdentifier();
+    m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), WTFMove(completionHandler), WTFMove(webPageProxyIdentifier) });
```
I think it'd look nicer if the completion handler came last.

cdumez replied:
No WTFMove() for webPageProxyIdentifier since this is a uint64_t.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
```diff
@@ -33,17 +34,15 @@
 
 private:
-    explicit WebPermissionController(WebPage&);
+    explicit WebPermissionController();
```
You can drop the explicit.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
```diff
@@ -62,6 +60,7 @@ class WebPermissionController final : public CanMakeWeakPtr<WebPermissionControl
         WebCore::ClientOrigin origin;
         WebCore::PermissionDescriptor descriptor;
         CompletionHandler<void(std::optional<WebCore::PermissionState>)> completionHandler;
+        WebPageProxyIdentifier identifier;
```
I think this should come before the completion handler.

RupinMittal replied:
Fixed!

cdumez commented on Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 46;
+	objectVersion = 52;
```
Please revert this change.

RupinMittal replied:
I'm curious, why did this change happen in the first place? (I didn't change it manually)

cdumez replied:
I don't know, Xcode...

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -104,11 +104,10 @@ void WebPermissionController::tryProcessingRequests()
 
         currentRequest.isWaitingForReply = true;
-        m_page->sendWithAsyncReply(Messages::WebPageProxy::QueryPermission(currentRequest.origin, currentRequest.descriptor), [this, weakThis = WeakPtr { *this }](auto state, bool shouldCache) {
+        WebProcess::singleton().sendWithAsyncReply(Messages::WebPermissionControllerProxy::Query(currentRequest.origin, currentRequest.descriptor, currentRequest.identifier), [this, weakThis = WeakPtr { *this }](auto state, bool shouldCache) {
```
Before sending message to UI process, you can check if page exists with:
`WebProcess::singleton().webPage(currentRequest.identifier)`, if not you may just invoke completionHandler of the request

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp:
```diff
@@ -0,0 +1,59 @@
+{
+    auto webPageProxy = m_process.webPage(identifier);
+    if (!webPageProxy)
+        completionHandler(WebCore::PermissionState::Prompt, false);
```
You should return here, otherwise, the next line will be executed

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp:
```diff
@@ -0,0 +1,59 @@
+#include "WebProcessProxy.h"
+#include <WebCore/ClientOrigin.h>
+#include <WebCore/PermissionDescriptor.h>
+#include <WebCore/PermissionState.h>
```
No need to include same headers in both .h and .cpp, and please use forward declaration if possible

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -26,33 +26,33 @@
 
-
 #include "WebPage.h"
+#include "WebPageProxyIdentifier.h"
```
No need to include the same header in .h and .cpp

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -88,7 +88,7 @@ void WebPermissionController::updateCache(const WebCore::ClientOrigin& origin, c
 void WebPermissionController::tryProcessingRequests()
 {
-    if (m_requests.isEmpty() || !m_page)
+    if (m_requests.isEmpty())
```
You can remove this, the while loop won't run if `m_requests` is empty

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -104,11 +104,10 @@ void WebPermissionController::tryProcessingRequests()
             RefPtr protectedThis { weakThis.get() };
             if (!protectedThis)
                 return;
-
```
Nit: no need to remove the blank line.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -26,33 +26,32 @@
         return completionHandler(cachedResult);
 
-    m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), WTFMove(completionHandler) });
+    auto pageIdentifier = WebPage::fromCorePage(page).identifier();
```
Why are you storing a PageIdentifier when you really need a WebPageProxyIdentifier?

We would probably call webPageProxyIdentifier() here.

RupinMittal replied:
We've gone back to using WebPageProxyIdentifier here. Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -103,8 +99,14 @@ void WebPermissionController::tryProcessingRequests()
             continue;
         }
 
+        auto* webPage = WebProcess::singleton().webPage(currentRequest.identifier);
```
I think Sihui recommended this but I disagree with the suggestion:
1. The request should store the WebPageProxyIdentifier, not the PageIdentifier
2. There is no point in looking up the WebPage here if you already have the right identifier you need to send over IPC
3. This adds more complexity to try and avoid an IPC round trip in a very edge case where the WebPage doesn't exist. This is just not worth it.
4. Added more WebPage-related code is not a good idea because we're going to support service workers, which won't have a WebPage.

szewai replied:
I don't think this is an edge case. In current impl, we block request handling when there's a pending request (waiting for UI process's reply). If a page makes multiple query requests and is gone, it's going to block requests from other pages for multiple rounds of IPCs.
I actually think we should store `std::optional<WeakPtr<WebPage>>` in the request, but PageIdentifier or `WeakPtr<WebPage>` might be enough for this patch.

cdumez replied:
Very edge case. Multiple requests in parallel is an edge case, multiple pages in a single WebProcess is already an edge case. Even if there were really several and some queueing were to happen, querying doesn't involve prompting and we can thus expect responses really fast.

Also, an IPC round trip is cheap.

cdumez replied:
I think you are adding complexity to try and optimize out a quick IPC round trip (quick since the UIProcess won't find the page and reply right away) for the case where:
1. There is more than one page in the WebProcess
and
2. Several of these WebPages are making query permission request in parallel
and
3. One of the WebPage gets closed before getting a response to its request(s)

RupinMittal replied:
We've gone back to using WebPageProxyIdentifier here. Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
```diff
@@ -61,6 +59,7 @@ class WebPermissionController final : public CanMakeWeakPtr<WebPermissionControl
     struct PermissionRequest {
         WebCore::ClientOrigin origin;
         WebCore::PermissionDescriptor descriptor;
+        WebCore::PageIdentifier identifier;
```
Why a PageIdentifier when we need a WebPageProxyIdentifier?

RupinMittal replied:
We've gone back to using WebPageProxyIdentifier here--fixed!

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -103,8 +99,14 @@ void WebPermissionController::tryProcessingRequests()
         }
 
+        auto* webPage = WebProcess::singleton().webPage(currentRequest.identifier);
+        if (!webPage) {
```
This check should go before `queryCache`

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -103,8 +99,14 @@ void WebPermissionController::tryProcessingRequests()
 
+        auto* webPage = WebProcess::singleton().webPage(currentRequest.identifier);
+        if (!webPage) {
+            auto takenRequest = m_requests.takeFirst();
+            takenRequest.completionHandler(WebCore::PermissionState::Prompt);
```
Can be written in one line: `m_requests.takeFirst().completionHandler(cachedResult);`

RupinMittal replied:
We've gone back to using WebPageProxyIdentifier here and not checking the existence of the page.

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -103,8 +99,14 @@ void WebPermissionController::tryProcessingRequests()
+        if (!webPage) {
+            auto takenRequest = m_requests.takeFirst();
+            takenRequest.completionHandler(WebCore::PermissionState::Prompt);
+        }
```
You need to call `continue` after completionHandler, otherwise you are going to send the message for this request.

RupinMittal replied:
We've gone back to using WebPageProxyIdentifier here and not checking the existence of the page.

cdumez reviewed (approved):
LGTM

szewai reviewed (approved):
Seems ready to land if bots are happy

RupinMittal committed:
Make WebPermissionController independent of the WebPage
https://bugs.webkit.org/show_bug.cgi?id=243597

Reviewed by Chris Dumez and Sihui Liu.

Currently, the Permissions API is not supported for workers. In order for it
to be supported, the Permissions API should function without the need for a
page. Previously, WebPermissionController was created and owned by WebPage.
This patch makes PermissionController a singleton that is created by the
WebProcess. So now, instead of Permissions::query() using the page to get the
WebPermissionController and send an IPC message to WebPageProxy to obtain
permissions states, it can access the WebPermissionController singleton to
send an IPC message that will eventually call the correct WebPageProxy. Note
that a page is still required to obtain the permission state, but this
refactoring is the first step in supporting the Permissions API for workers.

* Source/WebCore/Modules/permissions/PermissionController.cpp: Copied from Source/WebCore/Modules/permissions/Permissions.h.
(WebCore::sharedController):
(WebCore::PermissionController::shared):
(WebCore::PermissionController::setSharedController):
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/PermissionStatus.cpp:
(WebCore::PermissionStatus::PermissionStatus):
(WebCore::PermissionStatus::~PermissionStatus):
* Source/WebCore/Modules/permissions/PermissionStatus.h:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::Permissions):
(WebCore::Permissions::query):
* Source/WebCore/Modules/permissions/Permissions.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::permissionController): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::permissionController): Deleted.
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::permissionController): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp: Copied from Source/WebCore/Modules/permissions/Permissions.h.
(WebKit::WebPermissionControllerProxy::WebPermissionControllerProxy):
(WebKit::WebPermissionControllerProxy::~WebPermissionControllerProxy):
(WebKit::WebPermissionControllerProxy::query):
* Source/WebKit/UIProcess/WebPermissionControllerProxy.h: Copied from Source/WebCore/Modules/permissions/Permissions.h.
* Source/WebKit/UIProcess/WebPermissionControllerProxy.messages.in: Copied from Source/WebCore/Modules/permissions/Permissions.h.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::m_webPermissionController):
(WebKit::m_webLockRegistry): Deleted.
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::create):
(WebKit::WebPermissionController::query):
(WebKit::WebPermissionController::tryProcessingRequests):
(WebKit::WebPermissionController::WebPermissionController): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
* Source/WebKit/WebProcess/WebProcess.cpp:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):
* Source/WebKitLegacy/win/WebView.cpp:
(WebView::initWithFrame):

Canonical link: https://commits.webkit.org/253216@main

smfr commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -6717,6 +6720,12 @@
+		D952EC76289C4299006C240A /* WebPermissionControllerProxy.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = WebPermissionControllerProxy.h; sourceTree = "<group>"; };
+		D952EC77289C46B4006C240A /* WebPermissionControllerProxy.cpp */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.cpp.cpp; path = WebPermissionControllerProxy.cpp; sourceTree = "<group>"; };
+		D952EC79289C4AD0006C240A /* WebPermissionControllerProxy.messages.in */ = {isa = PBXFileReference; lastKnownFileType = text; path = WebPermissionControllerProxy.messages.in; sourceTree = "<group>"; };
+		D952EC7B289C9A4C006C240A /* WebPermissionControllerProxyMessages.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; name = WebPermissionControllerProxyMessages.h; path = ../Debug/DerivedSources/WebKit/WebPermissionControllerProxyMessages.h; sourceTree = BUILT_PRODUCTS_DIR; };
```
Bad paths here.
